### PR TITLE
nixosTest.nix-upgrade: init

### PR DESCRIPTION
**Pull request by tomberek**
Test out both nix upgrade-nix and a NixOS upgrade.

Inject a fake fallback-paths.nix assuming a stable -> latest upgrade.

The NixOS upgrade does not use nixos-rebuild switch due to the cost+annoyance of the instantiation needing
system.includeBuildDependencies.

@Mic92 

@RossComputerGuy (I removed the includeBuildDependencies and call the activation directly instead of using nixos-rebuild)

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/tests/nix/upgrade.nix [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+pkgs.testers.nixosTest rec {
+  name = "nix-upgrade-${pkgs.nixVersions.stable.version}-${pkgs.nixVersions.latest.version}";
+  meta.maintainers = with lib.maintainers; [ tomberek ];
+
+  nodes.machine =
+    { lib, modulesPath, ... }:
+    {
+      imports = [ nixos-module ];
+
+      virtualisation.memorySize = 9192;
```
Comment (RaitoBezarius) [inline nixos/tests/nix/upgrade.nix:55]: Hmmm, that seems a quite high memory size, is there a good reason for that?

File: nixos/tests/nix/upgrade.nix [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      virtualisation.cores = 4;
```
Comment (RaitoBezarius) [inline nixos/tests/nix/upgrade.nix:56]: Do you need that much cores or would you leave it to the defaults?

File: nixos/tests/nix/upgrade.nix [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      nix.package = pkgs.nixVersions.stable;
+      system.extraDependencies = [
+        fallback-paths-external
+        inputDrv.system
+      ];
+    };
+
+  testScript = ''
+    machine.start()
+    machine.succeed("udevadm settle")
```
Comment (RaitoBezarius) [inline nixos/tests/nix/upgrade.nix:67]: Why do you need `udevadm settle`?

**Review by RaitoBezarius — COMMENTED** [review]
Nice work here!

**Comment by RaitoBezarius** [comment]
> The NixOS upgrade does not use nixos-rebuild switch due to the cost+annoyance of the instantiation needing
system.includeBuildDependencies.

You could technically list statically the dependencies required like installer tests does it, if you really wanted to test it this way.

**Review by lf- — COMMENTED** [review]
can i encourage you to add some more templating and variables so it can support also checking cppnix -> lix (2.91) upgrades as well as perhaps some description of downgrades (the meat of which can be added later)? and maybe lix -> cppnix. if tests for this in both directions are in nixpkgs it should keep us all honest.

Currently this test hardcodes assumptions about which nix version is which in the fallback-paths thing for example. 

(note that we know lix 2.90 ships a broken upgrade-nix. you're also going to find that upgrade-nix (if the upgrade-nix implementation being run is from cppnix rather than lix) from lix *to* cppnix is probably going to fail due to an assumption in upgrade-nix which we since fixed in lix 2.91 caused by failure to consider that the nix pname could change)

File: nixos/tests/nix/upgrade.nix (hunk 2 above)
Comment (tomberek) [reply nixos/tests/nix/upgrade.nix:56]: Not really, it helped when I was iterating to do the full eval + build inside the QEMU. Removing...

**Comment by tomberek** [comment]
Replacing "nixVersions" is possible now.

**Comment by Mic92** [comment]
> > The NixOS upgrade does not use nixos-rebuild switch due to the cost+annoyance of the instantiation needing
> > system.includeBuildDependencies.
> 
> You could technically list statically the dependencies required like installer tests does it, if you really wanted to test it this way.

I did this in the past in other projects but it is a bit brittle because it can change over time depending on what you rebuild. It's ok if the installer tests do this, but here we don't need to test this part as well.

**Comment by Mic92** [comment]
@tomberek I made a few changes to your pull request i.e. using a non-privileged user for testing builds and referenced it in the nix package.

**Comment by roberth** [comment]
> a few changes

[diff](https://github.com/NixOS/nixpkgs/compare/c0f47f3e51537fc7e21d8fb9b407a1f4ed78227e..33b5aab88f6084ca14b3884ef533f54fdb043796?ws=1&w=1)

diff lgtm

**Comment by roberth** [comment]
> The NixOS upgrade does not use nixos-rebuild switch due to the cost+annoyance of the instantiation needing
> system.includeBuildDependencies.

Would switching to a NixOS specialisation be a viable option?
Those are pre-built, so they don't require builds to be done in the VM.

**Pull request merged**